### ISSUE TITLE
set_metadata() for newly created samples

### ISSUE THREAD

**Issue by andreas-wilm** [whitews/FlowKit#89]
**Is your feature request related to a problem? Please describe.**

There currently doesn't seem to be a way to set the metadata for newly created FCS files.

**Describe the solution you'd like**

I would like to read an input FCS file, massage the data, write an output file with the transferred input metadata. Another application is to create an FCS file from scratch and add some metadata.

**Describe alternatives you've considered**

I tried to set `sample.metadata` directly however the changes disappeared after `sample.export()`. I'm not sure if I did something wrong here:

```
>>> sample = fk.Sample("some.fcs")
>>> metadata = sample.metadata.copy()
>>> 'flowkit' in metadata
False
>>> metadata['flowkit'] = 'rocks'
>>> 'flowkit' in metadata
True

>>> df = sample.as_dataframe(source="raw")
>>> newsample = fk.Sample(df)
>>> newsample.metadata = metadata
>>> 'flowkit' in newsample.metadata
True
>>> newsample.export("test.fcs", source='raw')

>>> testsample = fk.Sample("test.fcs")
>>> 'flowkit' in testsample.metadata
False
```

**Additional context**

Congrats on the publication!

**Comment by whitews** [comment]
Thanks Andreas!

The issue here is that I need to make a clean pass-through function in FlowKit to FlowIO. It will be pretty easy to do, but I'm not sure of the best way. Would you prefer loading the dictionary and having it write the new metadata or provide a new dictionary of metadata when writing the new file? This gets more complicated if the user wants to specify metadata values such as the channel gain values. Perhaps we could allow some metadata values editable, while fixing some of the ones that are crucial to reading the accurate event values. Curious what your use case is and what your thoughts are?

**Comment by andreas-wilm** [comment]
Hey Scott,

in my specific case I need to be able to pass a few specific metadata key values pairs from the input file to the output file and I would also like to add one or two pairs. These are all non-standard values for annotation (data provenance etc.), i.e. I won't need to mess with any standard/fixed values.

I like your idea of having a set of immutable values and a set of editable user values. Something like `sample.metadata_user` or `sample.metadata_extra` in addition to the existing `sample.metadata` for example.

The simpler alternative might be to let the user mess directly with `sample.metadata` and trust it. This unfortunately would allow users to shoot themselves in the foot, so you have to assume that you only have responsible users :)

**Comment by lorenzgerber** [comment]
Hi guys... just some thoughts. In the most common R library to handle fcs data, metadata is split into two objects/dicts (parameters, description). Parameter describes meta data per channel. As a minimum there, the name would be required, numerical info per channel such as range etc can be calculated automatically at creation/save time. Further per channel info, such as description is passed on if available. The description object/dict contains freestyle name/value pairs. In this one the user is allowed to add/remove/change freely.

**Comment by whitews** [comment]
Hey guys, the latest commit on the develop branch should allow you to export new FCS files with metadata. Sorry this took so long but there were a few issues to take into account. The bulk of the changes happened in the FlowIO library, which I'm always careful about modifying.

I played around with a couple ways to do this and settled on what I felt was the most user-friendly. I went with the idea Louis floated, let users change the metadata dictionary directly. However, certain keywords will be ignored and set automatically to help ensure compliance with the FCS 3.1 specification. This felt like the most intuitive method.

If you could test it out and let me know if it works for you, that would be super helpful! Also, I'm toying with the idea to add an extra `metadata_blacklist` argument to the export function which would allow users to exclude certain values without having to pop or delete them from the original metadata. Thoughts on that would be appreciated as well.

**Comment by andreas-wilm** [comment]
Awesome. Just in time for us :) I ran into problem though, which might be just installation related.

I tried installing the develop branch version of flowkit (81146d0) via `python setup.py install` in a new conda environment, but got `error: The 'flowutils' distribution was not found and is required by FlowKit` (even though it's listed as dependency). No idea what's going on there.

So in another new conda environment, I first installed flowutils and flowio via pip (I saw that at least flowio is updated to 1.0.0) and then proceeded as above. Installation works, but the header is unchanged when I use the code example given in my original post :(
